### ISSUE TITLE
Implement DataFrameGroupBy, RollingGroupby, ExpandingGroupby

### ISSUE THREAD

**Issue by hermian** [ddelange/mapply#36]
#### What you were trying to do (and why)

The problem occurs when the apply function is called after applying groupby to the data frame.

#### What happened (including reproducible example)

<details><summary>Reproducible example</summary><p>

```py
def square(x):
    return x * x

df = pd.DataFrame({'A': 'a a b'.split(),
                   'B': [1,2,3],
                   'C': [4,6,5]})

df.groupby('A')[['B', 'C']].mapply(square)
```

```
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
~/workspace/git/screener/mmt/kospi_naver_db.py in <cell line: 1>()
----> 1 df.groupby('A', group_keys=True)[['B', 'C']].mapply(square)

/opt/homebrew/Caskroom/miniconda/base/envs/py38/lib/python3.8/site-packages/mapply/mapply.py in mapply(df_or_series, func, axis, n_workers, chunk_size, max_chunks_per_worker, progressbar, args, **kwargs)
     95 
     96     n_chunks = _choose_n_chunks(
---> 97         df_or_series.shape,
     98         opposite_axis,
     99         n_workers,

/opt/homebrew/Caskroom/miniconda/base/envs/py38/lib/python3.8/site-packages/pandas/core/groupby/groupby.py in __getattr__(self, attr)
    979             return self[attr]
    980 
--> 981         raise AttributeError(
    982             f"'{type(self).__name__}' object has no attribute '{attr}'"
    983         )

AttributeError: 'DataFrameGroupBy' object has no attribute 'shape'
```

</p></details>

**Comment by ddelange** [comment]
Hi @hermian 👋

Indeed, mapply currently only implements DataFrame and Series.

I think there are a number of edge cases where `DataFrameGroupBy.apply` produces different kinds of output (depending on keyword arguments, input shape, and function output shape).

For the naive approach, you can iterate over a DataFrameGroupBy to get df_or_series, and do mapply stuff with that:

```py
pd.concat(df.mapply() for df in data.groupby())
```

But as mentioned this only works for the 'obvious' cases where groupby() produces sub dataframes or series (and you might have to set axis on concat).

I think pandarallel has a complete implementation also for RollingDataFrameGroupBy etc. But last time I checked, there is only 1 chunk per worker and no way to configure that. So if one chunk takes much longer than the another chunks, in the end you'll have all cores (but one) idle, waiting for the last chunk to finish.

**Comment by ddelange** [comment]
DataFrameGroupBy done in #40

RollingGroupby and ExpandingGroupby have different mechanics [ref](https://github.com/pandas-dev/pandas/blob/v1.5.3/pandas/core/window/rolling.py#L606) and need another approach

**Comment by juanPabloMiceli** [comment]
About this, while using DataFrameGroupBy.mapply() I got:

```
mapply/_groupby.py:112: FutureWarning: DataFrameGroupBy.grouper is deprecated and will be removed in a future version of pandas.
  original_apply = getattr(df_or_series.grouper, attr)
mapply/_groupby.py:113: FutureWarning: DataFrameGroupBy.grouper is deprecated and will be removed in a future version of pandas.
  setattr(df_or_series.grouper, attr, MethodType(apply, df_or_series.grouper))
mapply/_groupby.py:117: FutureWarning: DataFrameGroupBy.grouper is deprecated and will be removed in a future version of pandas.
  setattr(df_or_series.grouper, attr, original_apply)
```
I am not adding an example because any DataFrameGroupBy.mapply() will throw those warnings.
Should I open a new issue?

**Comment by ddelange** [comment]
# Plan: RollingGroupby & ExpandingGroupby support for mapply

## Context

Issue [#36](https://github.com/ddelange/mapply/issues/36) requests support for `DataFrameGroupBy` (done in #40), `RollingGroupby`, and `ExpandingGroupby`. The pandas-v3 branch already refactored for pandas 3.0.1 compatibility. Unlike `GroupBy.apply` which has a patchable internal for-loop (`apply_groupwise`), the window groupby classes embed group logic deep in `GroupbyIndexer` bounds consumed by Cython functions — there is no single loop to patch. The approach is to decompose at the group level: split by groups, run non-grouped `Rolling`/`Expanding` per group in parallel via `multiprocessing_imap`, and reassemble with `pd.concat`.

The lazy generator consumption of `multiprocessing_imap` is exploited: groups are yielded lazily from a generator, fed to `pool.imap()` which consumes on-demand as workers free up, and results flow back one at a time via `yield from` for progressive assembly.

## Files to modify

| File | Action |
|------|--------|
| `src/mapply/_window_groupby.py` | **Create** — core `run_window_groupby_apply()` |
| `src/mapply/mapply.py` | **Edit** — detect `BaseWindowGroupby`, route to new handler |
| `src/mapply/__init__.py` | **Edit** — also patch `BaseWindowGroupby` in `init()` |
| `tests/test_mapply.py` | **Edit** — add RollingGroupby and ExpandingGroupby tests |

## Implementation

### 1. `src/mapply/_window_groupby.py` (new)

Follow the structure of `_groupby.py`. Single public function `run_window_groupby_apply`:

- Extract `_grouper.indices` and `_grouper.result_index` from the window groupby object
- Capture window params from the object's attributes:
  - **Rolling**: `window`, `min_periods`, `center`, `on`, `closed` (skip `step` — raises `NotImplementedError` for groupby; skip `win_type` — `.apply()` doesn't use weighted windows)
  - **Expanding**: `min_periods`
- Build a **lazy generator** yielding `(key, obj.iloc[indices[key]])` tuples — not materialized upfront
- Define `_process_group(key_and_data)` that creates a non-grouped `Rolling`/`Expanding` object on the group slice and calls `.apply(func, args=args, **kwargs)`
- Feed generator through `multiprocessing_imap` (lazy consumption by `pool.imap`, lazy result yield via `yield from`)
- Collect `(key, result)` pairs progressively in a `for` loop, append to `keys`/`parts` lists
- Handle empty groups: return empty result from an empty window `.apply()`
- Reassemble via `concat(parts, keys=keys, names=grouper.names + list(obj.index.names))`
- Handle `_as_index=False`: `result.reset_index(level=list(range(len(grouper.names))))`

Include the same BSD 3-Clause license header as `_groupby.py`.

```python
# ruff: noqa: ERA001
import logging
from collections.abc import Callable
from typing import Any

from mapply.parallel import multiprocessing_imap, tqdm

logger = logging.getLogger(__name__)


def run_window_groupby_apply(
    window_groupby: Any,
    func: Callable,
    *,
    n_workers: int,
    progressbar: bool,
    args: tuple[Any, ...] = (),
    **kwargs: Any,
):
    """Apply func to each group's window in parallel using multiprocessing_imap."""
    from pandas import concat
    from pandas.core.window.expanding import ExpandingGroupby

    grouper = window_groupby._grouper  # noqa: SLF001
    indices = grouper.indices
    result_index = grouper.result_index
    obj = window_groupby.obj
    as_index = window_groupby._as_index  # noqa: SLF001
    groupby_names = grouper.names

    is_expanding = isinstance(window_groupby, ExpandingGroupby)

    if is_expanding:
        window_kwargs = {
            "min_periods": window_groupby.min_periods,
        }
        window_method = "expanding"
    else:
        window_kwargs = {
            "window": window_groupby.window,
            "min_periods": window_groupby.min_periods,
            "center": window_groupby.center,
            "on": window_groupby.on,
            "closed": window_groupby.closed,
        }
        window_method = "rolling"

    # lazy generator: yield (key, group_slice) without materializing all groups
    def _group_iter():
        for key in result_index:
            yield key, obj.iloc[indices[key]]

    def _process_group(key_and_data):
        key, group_data = key_and_data
        window_obj = getattr(group_data, window_method)(**window_kwargs)
        result = window_obj.apply(func, args=args, **kwargs)
        return key, result

    # generator with length defined (for progressbar)
    groups = tqdm(_group_iter(), disable=True, total=len(result_index))
    processed = multiprocessing_imap(
        _process_group,
        groups,
        n_workers=n_workers,
        progressbar=progressbar,
    )

    # consume lazily from the multiprocessing_imap generator
    keys = []
    parts = []
    for key, part in processed:
        keys.append(key)
        parts.append(part)

    if not parts:
        empty_window = getattr(obj.iloc[:0], window_method)(**window_kwargs)
        return empty_window.apply(func, args=args, **kwargs)

    result = concat(parts, keys=keys, names=groupby_names + list(obj.index.names))

    if not as_index:
        result = result.reset_index(level=list(range(len(groupby_names))))

    return result
```

### 2. `src/mapply/mapply.py`

- Add import: `from mapply._window_groupby import run_window_groupby_apply`
- Before the existing `isinstance(df_or_series, GroupBy)` check (~line 124), add:
  ```python
  from pandas.core.window.rolling import BaseWindowGroupby
  if isinstance(df_or_series, BaseWindowGroupby):
      return run_window_groupby_apply(
          df_or_series, func, n_workers=n_workers,
          progressbar=progressbar, args=args, **kwargs,
      )
  ```
  `BaseWindowGroupby` is the common base for both `RollingGroupby` and `ExpandingGroupby`, and does NOT overlap with `GroupBy` (different class hierarchy).

### 3. `src/mapply/__init__.py`

- `BaseWindowGroupby` extends `BaseWindow(SelectionMixin)`, **not** `PandasObject`. The current `setattr(PandasObject, ...)` doesn't reach window objects.
- After the existing `setattr(PandasObject, apply_name, apply)` (line 97), add:
  ```python
  from pandas.core.window.rolling import BaseWindowGroupby
  setattr(BaseWindowGroupby, apply_name, apply)
  ```

### 4. `tests/test_mapply.py`

Add `test_rolling_groupby_mapply()` and `test_expanding_groupby_mapply()` following the existing pattern (`mapply.init(progressbar=False, chunk_size=1)`, compare `.apply()` vs `.mapply()` with `pd.testing.assert_frame_equal`/`assert_series_equal`).

```python
def test_rolling_groupby_mapply():
    """Assert RollingGroupby behaviour is equivalent."""
    mapply.init(progressbar=False, chunk_size=1)

    np.random.seed(42)  # noqa: NPY002
    df = pd.DataFrame(
        {
            "A": np.random.randint(0, 100, 200),  # noqa: NPY002
            "B": np.random.randint(0, 100, 200),  # noqa: NPY002
            "group": [0] * 100 + [1] * 100,
        }
    )

    # basic RollingGroupby with custom func
    pd.testing.assert_frame_equal(
        df.groupby("group").rolling(3).apply(lambda x: x.sum()),
        df.groupby("group").rolling(3).mapply(lambda x: x.sum()),
    )

    # min_periods
    pd.testing.assert_frame_equal(
        df.groupby("group").rolling(5, min_periods=2).apply(lambda x: x.mean()),
        df.groupby("group").rolling(5, min_periods=2).mapply(lambda x: x.mean()),
    )

    # center=True
    pd.testing.assert_frame_equal(
        df.groupby("group").rolling(3, center=True).apply(lambda x: x.max()),
        df.groupby("group").rolling(3, center=True).mapply(lambda x: x.max()),
    )

    # column selection (Series result)
    pd.testing.assert_series_equal(
        df.groupby("group")["A"].rolling(3).apply(lambda x: x.sum()),
        df.groupby("group")["A"].rolling(3).mapply(lambda x: x.sum()),
    )

    # as_index=False
    pd.testing.assert_frame_equal(
        df.groupby("group", as_index=False).rolling(3).apply(lambda x: x.sum()),
        df.groupby("group", as_index=False).rolling(3).mapply(lambda x: x.sum()),
    )

    # multi-level groupby
    df["group2"] = list(range(2)) * 100
    pd.testing.assert_frame_equal(
        df.groupby(["group", "group2"]).rolling(3).apply(lambda x: x.sum()),
        df.groupby(["group", "group2"]).rolling(3).mapply(lambda x: x.sum()),
    )

    # time-based rolling with 'on' parameter
    df_ts = pd.DataFrame(
        {
            "A": np.random.randint(0, 100, 200),  # noqa: NPY002
            "dt": pd.date_range("2020-01-01", periods=200, freq="D"),
            "group": [0] * 100 + [1] * 100,
        }
    )
    pd.testing.assert_frame_equal(
        df_ts.groupby("group").rolling("3D", on="dt").apply(lambda x: x.sum()),
        df_ts.groupby("group").rolling("3D", on="dt").mapply(lambda x: x.sum()),
    )

    # empty groupby
    pd.testing.assert_frame_equal(
        df.iloc[:0].groupby("group").rolling(3).apply(lambda x: x.sum()),
        df.iloc[:0].groupby("group").rolling(3).mapply(lambda x: x.sum()),
    )

    # n_workers=1 (single-process fallback, no pool spawned)
    mapply.init(progressbar=False, chunk_size=1, n_workers=1)
    pd.testing.assert_frame_equal(
        df.groupby("group").rolling(3).apply(lambda x: x.sum()),
        df.groupby("group").rolling(3).mapply(lambda x: x.sum()),
    )


def test_expanding_groupby_mapply():
    """Assert ExpandingGroupby behaviour is equivalent."""
    mapply.init(progressbar=False, chunk_size=1)

    np.random.seed(42)  # noqa: NPY002
    df = pd.DataFrame(
        {
            "A": np.random.randint(0, 100, 200),  # noqa: NPY002
            "B": np.random.randint(0, 100, 200),  # noqa: NPY002
            "group": [0] * 100 + [1] * 100,
        }
    )

    # basic ExpandingGroupby with custom func
    pd.testing.assert_frame_equal(
        df.groupby("group").expanding().apply(lambda x: x.sum()),
        df.groupby("group").expanding().mapply(lambda x: x.sum()),
    )

    # min_periods
    pd.testing.assert_frame_equal(
        df.groupby("group").expanding(min_periods=3).apply(lambda x: x.mean()),
        df.groupby("group").expanding(min_periods=3).mapply(lambda x: x.mean()),
    )

    # column selection (Series result)
    pd.testing.assert_series_equal(
        df.groupby("group")["A"].expanding().apply(lambda x: x.sum()),
        df.groupby("group")["A"].expanding().mapply(lambda x: x.sum()),
    )

    # as_index=False
    pd.testing.assert_frame_equal(
        df.groupby("group", as_index=False).expanding().apply(lambda x: x.sum()),
        df.groupby("group", as_index=False).expanding().mapply(lambda x: x.sum()),
    )

    # multi-level groupby
    df["group2"] = list(range(2)) * 100
    pd.testing.assert_frame_equal(
        df.groupby(["group", "group2"]).expanding().apply(lambda x: x.sum()),
        df.groupby(["group", "group2"]).expanding().mapply(lambda x: x.sum()),
    )

    # empty groupby
    pd.testing.assert_frame_equal(
        df.iloc[:0].groupby("group").expanding().apply(lambda x: x.sum()),
        df.iloc[:0].groupby("group").expanding().mapply(lambda x: x.sum()),
    )

    # n_workers=1 (single-process fallback)
    mapply.init(progressbar=False, chunk_size=1, n_workers=1)
    pd.testing.assert_frame_equal(
        df.groupby("group").expanding().apply(lambda x: x.sum()),
        df.groupby("group").expanding().mapply(lambda x: x.sum()),
    )
```

## Verification

1. `pytest tests/` — all existing + new tests pass
2. `ruff check src/ tests/` — no lint violations
3. Manual smoke test:
   ```python
   import pandas as pd, numpy as np, mapply
   mapply.init(n_workers=2, progressbar=True, chunk_size=1)
   df = pd.DataFrame({"A": np.random.randn(1000), "g": [0]*500 + [1]*500})
   # Should show tqdm progress and match native output:
   df.groupby("g").rolling(10).mapply(lambda x: x.sum())
   df.groupby("g").expanding().mapply(lambda x: x.sum())
   ```

**Comment by ddelange** [comment]
Released [0.2.0](https://github.com/ddelange/mapply/releases/tag/0.2.0)
